File: evals/refs/mr.zig
const std = @import("std");

fn mulMod(a: u128, b: u128, m: u128) u128 {
    return (a * b) % m;
}

fn powMod(base: u128, exp: u128, m: u128) u128 {
    var result: u128 = 1;
    var b = base % m;
    var e = exp;
    while (e > 0) {
        if ((e & 1) == 1) result = mulMod(result, b, m);
        b = mulMod(b, b, m);
        e >>= 1;
    }
    return result;
}

fn millerRabin(n: u64) bool {
    if (n < 2) return false;
    const small_primes = [_]u64{ 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37 };
    for (small_primes) |p| {
        if (n == p) return true;
        if (n % p == 0) return false;
    }
    var d: u64 = n - 1;
    var s: u64 = 0;
    while ((d & 1) == 0) {
        d >>= 1;
        s += 1;
    }
    const N: u128 = n;
    for (small_primes) |a| {
        var x = powMod(a, d, N);
        if (x == 1 or x == N - 1) continue;
        var composite = true;
        var i: u64 = 0;
        while (i + 1 < s) : (i += 1) {
            x = mulMod(x, x, N);
            if (x == N - 1) {
                composite = false;
                break;
            }
        }
        if (composite) return false;
    }
    return true;
}

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();

    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;

    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);

    var out_buf: [4096]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;

    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var case: usize = 0;
    while (case < t) : (case += 1) {
        const n = try std.fmt.parseInt(u64, it.next().?, 10);
        try w.writeAll(if (millerRabin(n)) "true\n" else "false\n");
    }
    try w.flush();
}

File: evals/refs/mr.rs
use std::io::{self, Read, Write};

fn mulmod(a: u64, b: u64, m: u64) -> u64 {
    ((a as u128 * b as u128) % m as u128) as u64
}

fn powmod(mut base: u64, mut exp: u64, m: u64) -> u64 {
    let mut res: u64 = 1;
    base %= m;
    while exp > 0 {
        if exp & 1 == 1 {
            res = mulmod(res, base, m);
        }
        base = mulmod(base, base, m);
        exp >>= 1;
    }
    res
}

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    let primes = [2u64, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37];
    for &p in &primes {
        if n == p {
            return true;
        }
        if n % p == 0 {
            return false;
        }
    }
    let mut d = n - 1;
    let mut s = 0;
    while d & 1 == 0 {
        d >>= 1;
        s += 1;
    }
    'witness: for &a in &primes {
        let mut x = powmod(a, d, n);
        if x == 1 || x == n - 1 {
            continue;
        }
        for _ in 0..s - 1 {
            x = mulmod(x, x, n);
            if x == n - 1 {
                continue 'witness;
            }
        }
        return false;
    }
    true
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut nums = input.split_whitespace();
    let t: usize = nums.next().unwrap().parse().unwrap();
    let mut out = String::new();
    for _ in 0..t {
        let n: u64 = nums.next().unwrap().parse().unwrap();
        out.push_str(if is_prime(n) { "true\n" } else { "false\n" });
    }
    io::stdout().write_all(out.as_bytes()).unwrap();
}
